File: typescript/Behavioral/22-Visitor/Structure/ProjectJobList.ts
import type { IJobElement } from "../Element/IJobElement";
import type { IJobVisitor } from "../Visitor/IJobVisitor";

export class ProjectJobList {
  private jobs: IJobElement[] = [];

  add(job: IJobElement): void {
    this.jobs.push(job);
  }

  // One traversal method serves every visitor
  accept(visitor: IJobVisitor): void {
    for (const job of this.jobs) job.accept(visitor);
  }
}

File: csharp/Behavioral/22-Visitor/Structure/ProjectJobList.cs
namespace Visitor;

public class ProjectJobList
{
    private readonly List<IJobElement> _jobs = new();

    public void Add(IJobElement job) => _jobs.Add(job);

    // One traversal method serves every visitor
    public void Accept(IJobVisitor visitor)
    {
        foreach (var job in _jobs)
            job.Accept(visitor);
    }
}
